src/mbedtls.c

```diff
@@ -730,4 +730,513 @@ _libssh2_dh_dtor(_libssh2_dh_ctx *dhctx)
+        goto failed;
+
+    if(mbedtls_mpi_read_binary(&(*ctx)->d, exponent, exponentlen) != 0)
+        goto failed;
+
+    if(mbedtls_ecp_mul(&(*ctx)->grp, &(*ctx)->Q,
+                       &(*ctx)->d, &(*ctx)->grp.G,
+                       mbedtls_ctr_drbg_random,
+                       &_libssh2_mbedtls_ctr_drbg) != 0)
+        goto failed;
+
+    if(mbedtls_ecp_check_privkey(&(*ctx)->grp, &(*ctx)->d) == 0)
+        goto cleanup;
+
+failed:
+
+    _libssh2_mbedtls_ecdsa_free(*ctx);
+    *ctx = NULL;
+
+cleanup:
+
+    if(decrypted) {
+        _libssh2_string_buf_free(session, decrypted);
+    }
+
+    return (*ctx == NULL) ? -1 : 0;
+}
+
+/* _libssh2_ecdsa_new_private
+ *
+ * Creates a new private key given a file path and password
+ *
+ */
+
+int
+_libssh2_mbedtls_ecdsa_new_private(libssh2_ecdsa_ctx **ctx,
+                                   LIBSSH2_SESSION *session,
+                                   const char *filename,
+                                   const unsigned char *pwd)
+{
+    mbedtls_pk_context pkey;
+    unsigned char *data;
+    size_t data_len;
+
+    if(mbedtls_pk_load_file(filename, &data, &data_len) != 0)
+        goto cleanup;
+
+    mbedtls_pk_init(&pkey);
+
+    if(_libssh2_mbedtls_parse_eckey(ctx, &pkey, session,
+                                    data, data_len, pwd) == 0)
+        goto cleanup;
+
+    _libssh2_mbedtls_parse_openssh_key(ctx, session, data, data_len, pwd);
+
+cleanup:
+
+    mbedtls_pk_free(&pkey);
+
+    _libssh2_mbedtls_safe_free(data, data_len);
+
+    return (*ctx == NULL) ? -1 : 0;
+}
+
+/* _libssh2_ecdsa_new_private
+ *
+ * Creates a new private key given a file data and password
+ *
+ */
+
+int
+_libssh2_mbedtls_ecdsa_new_private_frommemory(libssh2_ecdsa_ctx **ctx,
+                                              LIBSSH2_SESSION *session,
+                                              const char *data,
+                                              size_t data_len,
+                                              const unsigned char *pwd)
+{
+    unsigned char *ntdata;
+    mbedtls_pk_context pkey;
+
+    mbedtls_pk_init(&pkey);
+
+    ntdata = LIBSSH2_ALLOC(session, data_len + 1);
+
+    if(ntdata == NULL)
+        goto cleanup;
+
+    memcpy(ntdata, data, data_len);
+
+    if(_libssh2_mbedtls_parse_eckey(ctx, &pkey, session,
+                                    ntdata, data_len + 1, pwd) == 0)
+        goto cleanup;
+
+    _libssh2_mbedtls_parse_openssh_key(ctx, session,
+                                       ntdata, data_len + 1, pwd);
+
+cleanup:
+
+    mbedtls_pk_free(&pkey);
+
+    _libssh2_mbedtls_safe_free(ntdata, data_len);
+
+    return (*ctx == NULL) ? -1 : 0;
+}
+
+static unsigned char *
+_libssh2_mbedtls_mpi_write_binary(unsigned char *buf,
+                                  const mbedtls_mpi *mpi,
+                                  size_t bytes)
+{
+    unsigned char *p = buf;
+
+    p += 4;
+    *p = 0;
+
+    mbedtls_mpi_write_binary(mpi, p + 1, bytes - 1);
```

Comment:
also advancing p here, and if bytes is 0 length this could go negative. Same below in the memmove.